File: src/main/scala/org/tindalos/principle/domain/analyzers/structure/PackageCohesionAnalyzer.scala
package org.tindalos.principle.domain.analyzers.structure

import org.tindalos.principle.domain.AnalysisInput
import org.tindalos.principle.domain.analyzers.Analyzer
import org.tindalos.principle.domain.analyzers.structure.PackageCohesionModule.PackageName
import org.tindalos.principle.domain.analyzers.structure.GroupingResult
import org.tindalos.principle.domain.constraints.Constraints
import scala.collection.JavaConverters._

class PackageCohesionAnalyzer(buildComponents:(PackageName, Set[Node]) => Set[(PackageName, NodeGroup)]
                              , makeStructureHints1: Set[Node] => GroupingResult
                              , findDetachableSubgraphs: Set[Node] => SubgraphDecomposition
                              , collapseToLimit: Set[NodeGroup] => Set[NodeGroup]) extends Analyzer {
    
    override def analyze(input: AnalysisInput) = {

      val packagesWithCohesions = buildComponents(input.analysisPlan.basePackage, input.nodes) 
      val structureHints1 = makeStructureHints1(input.nodes)
      val structureHints2 = findDetachableSubgraphs(input.nodes)

      val cohesiveGroups: Option[Set[NodeGroup]] =
        if (input.packageCouplingExpectations().isPresent) {
          val initialGroups = input.nodes.map(n => new NodeGroup(java.util.Collections.singleton(n)))
          Some(collapseToLimit(initialGroups))
        }
        else None

      val javaPackages = packagesWithCohesions.toMap.asJava
      val javaGroups: java.util.Optional[java.util.Set[NodeGroup]] = cohesiveGroups match {
        case Some(groups) => java.util.Optional.of(groups.asJava)
        case None         => java.util.Optional.empty()
      }

      new CohesionAnalysisResult(javaPackages, javaGroups, structureHints1, structureHints2)
    }

    override def isEnabled(expectations: Constraints) =
      expectations.packageCoupling().isPresent && expectations.packageCoupling().get().grouping().isPresent

}

File: src/main/java/org/tindalos/principle/domain/analyzers/structure/PackageCohesionAnalyzer.java
package org.tindalos.principle.domain.analyzers.structure;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

import org.tindalos.principle.domain.plan.AnalysisInput;
import org.tindalos.principle.domain.analyzers.Analyzer;
import org.tindalos.principle.domain.constraints.Constraints;

/**
 * Analyzer producing package cohesion metrics and structure hints.
 */
public final class PackageCohesionAnalyzer implements Analyzer {

    @Override
    public CohesionAnalysisResult analyze(AnalysisInput input) {
        var nodes = input.nodes();

        var packagesWithCohesions = PackageCohesionModule.componentsFromPackages(input.analysisPlan().basePackage(), nodes);
        var structureHints1 = PackageStructureHints1Finder.makeGroups(nodes);
        var structureHints2 = Graph.findDetachableSubgraphs(nodes);

        Optional<Set<NodeGroup>> cohesiveGroups;
        if (input.packageCouplingExpectations().isPresent()) {
            var initialGroups = nodes.stream()
                    .map(node -> new NodeGroup(Set.of(node)))
                    .collect(Collectors.toUnmodifiableSet());
            cohesiveGroups = Optional.of(CohesiveGroupsDiscoveryModule.collapseToLimit(initialGroups));
        } else {
            cohesiveGroups = Optional.empty();
        }

        Map<String, NodeGroup> javaPackages = new LinkedHashMap<>();
        packagesWithCohesions.forEach(entry -> javaPackages.put(entry.getKey(), entry.getValue()));

        return new CohesionAnalysisResult(javaPackages, cohesiveGroups, structureHints1, structureHints2);
    }

    @Override
    public boolean isEnabled(Constraints expectations) {
        return expectations.packageCoupling().isPresent()
                && expectations.packageCoupling().get().grouping().isPresent();
    }
}